codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpRpcProtocolGenerator.java

```diff
@@ -0,0 +1,329 @@
+     */
+    public HttpRpcProtocolGenerator(boolean isErrorCodeInBody) {
+        this.isErrorCodeInBody = isErrorCodeInBody;
+    }
+
+    @Override
+    public ApplicationProtocol getApplicationProtocol() {
+        return ApplicationProtocol.createDefaultHttpApplicationProtocol();
+    }
+
+    /**
+     * Gets the content-type for a request body.
+     *
+     * @return Returns the default content-type.
+     */
+    protected abstract String getDocumentContentType();
+
+    @Override
+    public void generateSharedSerializerComponents(GenerationContext context) {
+        serializingDocumentShapes.addAll(ProtocolUtils.resolveRequiredDocumentShapeSerde(
+                context.getModel(), serializingDocumentShapes));
+        generateDocumentBodyShapeSerializers(context, serializingDocumentShapes);
+    }
+
+    /**
+     * Generates serialization functions for shapes in the passed set. These functions
+     * should return a value that can then be serialized by the implementation of
+     * {@code serializeInputDocument}.
+     *
+     * @param context The generation context.
+     * @param shapes  The shapes to generate serialization for.
+     */
+    protected abstract void generateDocumentBodyShapeSerializers(GenerationContext context, Set<Shape> shapes);
+
+    @Override
+    public void generateRequestSerializers(GenerationContext context) {
+        TopDownIndex topDownIndex = context.getModel().getKnowledge(TopDownIndex.class);
+
+        Set<OperationShape> containedOperations = new TreeSet<>(
+                topDownIndex.getContainedOperations(context.getService()));
+        for (OperationShape operation : containedOperations) {
+            generateOperationSerializer(context, operation);
+        }
+    }
+
+    private void generateOperationSerializer(GenerationContext context, OperationShape operation) {
+        GoStackStepMiddlewareGenerator middleware = GoStackStepMiddlewareGenerator.createSerializeStepMiddleware(
+                ProtocolGenerator.getSerializeMiddlewareName(operation.getId(), getProtocolName()),
+                ProtocolUtils.OPERATION_SERIALIZER_MIDDLEWARE_ID);
+
+        SymbolProvider symbolProvider = context.getSymbolProvider();
+        Model model = context.getModel();
+        Shape inputShape = ProtocolUtils.expectInput(model, operation);
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+        ApplicationProtocol applicationProtocol = getApplicationProtocol();
+        Symbol requestType = applicationProtocol.getRequestType();
+        GoWriter writer = context.getWriter();
+
+        middleware.writeMiddleware(context.getWriter(), (generator, w) -> {
+            writer.addUseImports(SmithyGoDependency.SMITHY);
+            writer.addUseImports(SmithyGoDependency.FMT);
+            writer.addUseImports(SmithyGoDependency.SMITHY_HTTP_BINDING);
+
+            // Cast the input request to the transport request type and check for errors.
```

Comment:
It might be useful to [generate a single function](https://github.com/awslabs/smithy-typescript/blob/master/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpRpcProtocolGenerator.java#L98:L129) that handles preparing the request and [just invoking that with the altered contents](https://github.com/awslabs/smithy-typescript/blob/master/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpRpcProtocolGenerator.java#L182:L186). A lot of this content looks static enough to warrant it, and there'll be a lot of duplicated code other wise.

Reply:
Hmmm maybe. I can probably make it work for RPC but I'm not sure about rest since the encoder needs to be passed around. I will say that duplicated code doesn't matter *that* much in the generated go, where the binary size is less important than performance.

Reply:
Looked into this a bit more. Even for RPC this would require a pretty large overhaul of the way the http binding encoder works. Probably worth doing, so I'll add a comment about it, but I think it would be out of scope.